bundles/org.eclipse.equinox.common/src/org/eclipse/core/runtime/ServiceCaller.java

```diff
@@ -140,6 +146,123 @@ public static <Service> boolean callOnce(Class<?> caller, Class<Service> service
 		return new ServiceCaller<>(caller, serviceType, filter).getCallUnget(consumer);
 	}
 
+	/**
+	 * Not thread save.
+	 * <p>
+	 * Service registrations are obtained upon creation, service instances are get
+	 * on the first request (i.e. the first instance when {@link #isAvailable()} or
+	 * {@link #get()} is called. Services that are unregistered in the meantime are
+	 * discarded and newly registered services are ignored.
+	 * </p>
+	 * 
+	 * @param <S> the service type for this caller
+	 * @since 3.17
+	 */
+	public interface ServiceUsage<S> extends AutoCloseable, Supplier<S> {
+
+		/**
+		 * Returns true if any service is available.
+		 * 
+		 * @return true if any instance is available.
+		 */
+		boolean isAvailable();
+
+		/**
+		 * Gets the first available service implementation.
+		 */
+		@Override
+		S get() throws NoSuchElementException;
```

Comment:
I would suggest to use an optional here, thats much more convenient and flexible and one can use the "orThrows" methods to give meaningful exception/error messages.

Reply:
I thought about that as well and actually wanted `ServiceUsage` to be an `Optional`, but since `Optional` is final this is not feasible (even if `ServiceUsage` would not be an interface).
Therefore my goal was to mimic Optional here with a similar `get()` method and isAvailable() (which also can be renamed to `isPresent()`).
But making it a Supplier that returns an Optional seems to be a bit inconvenient for me, because to eventually get the service instance one has to call `serviceUsage.get().get()` (if the service must be there), which I don't find nice.


> But if, please don't use Exceptions to signal missing services, this gives very very bad error reporting, better use Optionals so the caller of that method can give good errors or even use defaults if a service is missing.

Is there another default that could be used than `null`? I think returning null is not desired, for the same reasons why one should use Optional over null. Furthermore I think the error message that is thrown by the `ServiceUsage` implementation is more descriptive than the one of `Optional` because it explicitly mentions the unavailable service.

Reply:
If you can return null or throw exception you can return an optional. Calling get() on `Optional` is actually not recommended, you should use [orElse](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElse-T-) or [orElseThrow](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElseThrow-java.util.function.Supplier-) to either handle or signal a missing item. Also the method name "get" is jsut the choice of supplier, as you implement an interface here you can always choose a better name like `service` or `fetch` as a method name.